File: squirrel-backend/models/rss.py
from datetime import datetime
from typing import Optional

from sqlalchemy import Boolean, DateTime, Index, Integer, JSON, Text, UniqueConstraint, VARCHAR
from sqlalchemy.orm import Mapped, mapped_column

from models import Base
from models.mixins.serializer import SerializerMixin


class RssAccount(Base, SerializerMixin):
    __tablename__ = 'rss_account'

    __table_args__ = (
        UniqueConstraint('user_id', 'provider', 'base_url', 'name', name='uix_rss_account_user_provider_url_name'),
        Index('ix_rss_account_user_provider', 'user_id', 'provider'),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    user_id: Mapped[int] = mapped_column(Integer, nullable=False)
    provider: Mapped[str] = mapped_column(VARCHAR(32), nullable=False)
    name: Mapped[str] = mapped_column(VARCHAR(120), nullable=False)
    base_url: Mapped[str] = mapped_column(VARCHAR(2048), nullable=False)
    username: Mapped[Optional[str]] = mapped_column(VARCHAR(255), nullable=True)
    credential_encrypted: Mapped[str] = mapped_column(Text, nullable=False)
    enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    sync_entry_limit: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    is_deleted: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    last_sync_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    last_error: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(default=lambda: datetime.now())
    updated_at: Mapped[datetime] = mapped_column(default=lambda: datetime.now(), onupdate=lambda: datetime.now())


class RssFeed(Base, SerializerMixin):
    __tablename__ = 'rss_feed'

    __table_args__ = (
        UniqueConstraint('account_id', 'external_feed_id', name='uix_rss_feed_account_external'),
        Index('ix_rss_feed_account_enabled', 'account_id', 'enabled'),
        Index('ix_rss_feed_user_title', 'user_id', 'title'),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    user_id: Mapped[int] = mapped_column(Integer, nullable=False)
    account_id: Mapped[int] = mapped_column(Integer, nullable=False)
    external_feed_id: Mapped[str] = mapped_column(Text, nullable=False)
    title: Mapped[str] = mapped_column(VARCHAR(512), nullable=False)
    feed_url: Mapped[Optional[str]] = mapped_column(VARCHAR(2048), nullable=True)
    site_url: Mapped[Optional[str]] = mapped_column(VARCHAR(2048), nullable=True)
    icon_url: Mapped[Optional[str]] = mapped_column(VARCHAR(2048), nullable=True)
    category: Mapped[Optional[str]] = mapped_column(VARCHAR(255), nullable=True)
    enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    raw_data: Mapped[Optional[dict]] = mapped_column(JSON, nullable=True)
    last_entry_sync_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(default=lambda: datetime.now())
    updated_at: Mapped[datetime] = mapped_column(default=lambda: datetime.now(), onupdate=lambda: datetime.now())


class RssEntry(Base, SerializerMixin):
    __tablename__ = 'rss_entry'

    __table_args__ = (
        UniqueConstraint('feed_id', 'external_entry_id', name='uix_rss_entry_feed_external'),
        Index('ix_rss_entry_feed_published', 'feed_id', 'published_at'),
        Index('ix_rss_entry_user_published', 'user_id', 'published_at'),
        Index('ix_rss_entry_user_read', 'user_id', 'is_read', 'published_at'),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    user_id: Mapped[int] = mapped_column(Integer, nullable=False)
    account_id: Mapped[int] = mapped_column(Integer, nullable=False)
    feed_id: Mapped[int] = mapped_column(Integer, nullable=False)
    external_entry_id: Mapped[str] = mapped_column(Text, nullable=False)
    canonical_url: Mapped[str] = mapped_column(VARCHAR(2048), nullable=False)
    title: Mapped[str] = mapped_column(Text, nullable=False)
    summary: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    thumbnail: Mapped[Optional[str]] = mapped_column(VARCHAR(2048), nullable=True)
    author: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    published_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    is_read: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    is_starred: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    raw_data: Mapped[Optional[dict]] = mapped_column(JSON, nullable=True)
    created_at: Mapped[datetime] = mapped_column(default=lambda: datetime.now())
    updated_at: Mapped[datetime] = mapped_column(default=lambda: datetime.now(), onupdate=lambda: datetime.now())

File: squirrel-frontend/src/features/rss/types/rss.ts
// DTOs for /api/rss/* — field shapes mirror the backend serializers.
// See backend: domains/rss/application/services/{account,library}/serialization.py
//
// Notes:
// - Datetimes here use .isoformat() ("2026-06-18T20:23:00"), unlike the
//   SerializerMixin "YYYY-MM-DD HH:MM:SS" used elsewhere — both are TS string.
// - credential_encrypted / is_deleted / raw_data / user_id are STRIPPED from
//   responses; never re-add credential fields to these types.

export interface RssAccount {
  id: number
  provider: string
  name: string
  base_url: string
  username: string | null
  enabled: boolean
  sync_entry_limit: number | null
  last_sync_at: string | null
  last_error: string | null
  created_at: string | null
  updated_at: string | null
}

export interface RssFeed {
  id: number
  account_id: number
  external_feed_id: string
  title: string
  feed_url: string | null
  site_url: string | null
  icon_url: string | null
  category: string | null
  enabled: boolean
  open_method: string | null
  last_entry_sync_at: string | null
}

export interface RssEntry {
  id: number
  account_id: number
  feed_id: number
  external_entry_id: string
  canonical_url: string
  title: string
  summary: string | null
  thumbnail: string | null
  author: string | null
  published_at: string | null
  is_read: boolean
  is_starred: boolean
}

/** GET /api/rss/entries/recently-viewed item. */
export interface RssEntryRecentlyViewed extends RssEntry {
  viewed_at: string
}

/** Wrapper for GET /api/rss/entries. */
export interface RssEntryListResponse {
  total: number
  page: number
  pageSize: number
  data: RssEntry[]
}

/** Wrapper for GET /api/rss/feeds and /api/rss/accounts. */
export interface RssListResponse<T> {
  data: T[]
}

/** syncRssFeed returns the count of newly synced entries. */
export interface RssFeedSyncResult {
  entries?: number
}

/** GET /api/rss/accounts/:id/sync-status payload. */
export interface RssSyncStatus {
  running?: boolean
  sync_mode?: string
  phase?: string
  entries_fetched?: number | null
  entries_synced?: number | null
  feeds_synced?: number | null
  message?: string | null
  error?: string | null
}

/** POST /api/rss/entries/bulk result. */
export interface RssBulkUpdateResult {
  updated: number
}

/** Generic ack shape for RSS mutations with no meaningful body. */
export interface RssMutationResult {
  [key: string]: unknown
}

/** POST /api/rss/accounts/:id/sync/start result. */
export interface RssSyncStartResult {
  status?: string
  account_id?: string | number
}

/** POST /api/rss/accounts/test result. */
export interface RssAccountTestResult {
  ok?: boolean
  feed_count?: number
  message?: string
  error?: string
}

